Handle exception on integer conversion of volume size.

govint opened the pull request:
Added a try-except block to catch ValueErrors when converting volume size.

Verified as below:
docker volume create --driver=vsphere --name=y_7 -o size=100geeebEEE
Error response from daemon: create y_7: VolumeDriver.Create: Invalid format for size.
Valid sizes must be of form X[mMgGtT]b where X is aninteger. Default = 100mb

govint commented:
No, this validation is needed in the service code, all of the option validations are done in the service only as it knows the limits and what values are allowed for each supported option.

Even if the plugin is enhacned to do options validation, the service code can't rely that the plugin has provided correct options, it needs to handle this exception here.

shaominchen commented on esx_service/utils/convert.py:
```diff
@@ -24,7 +24,11 @@ def convert_to_MB(vol_size_str):
 
     unit = vol_size_str[-2:].upper()
-    value = int(vol_size_str[:-2])
+    try:
```
The current fix makes sense as it's consistent with the logic at line 41 - logging an error and returning 0. However, I checked the caller's logic in auth.py - we are not checking the invalid return value 0 at all. Ideally we should fix that as well to make it more robust.

govint replied:
Isn't this function supposed to return result once its set?

def check_privileges_for_command(cmd, opts, tenant_uuid, datastore_url, privileges):
    """
        Check whether the (tenant_uuid, datastore) has the privileges to run
        the given command.

    """
    result = None
    if not privileges:
        result = error_code_to_message[ErrorCode.PRIVILEGE_NO_PRIVILEGE]
        return result

    cmd_need_mount_privilege = [CMD_ATTACH, CMD_DETACH]
    if cmd in cmd_need_mount_privilege:
  ......

    if cmd == CMD_CREATE:
        if not has_privilege(privileges, auth_data_const.COL_ALLOW_CREATE):
            result = error_code_to_message[ErrorCode.PRIVILEGE_NO_CREATE_PRIVILEGE] <---
        if not check_max_volume_size(opts, privileges):
            result = error_code_to_message[ErrorCode.PRIVILEGE_MAX_VOL_EXCEED] <----
        if not check_usage_quota(opts, tenant_uuid, datastore_url, privileges):
            result = error_code_to_message[ErrorCode.PRIVILEGE_USAGE_QUOTA_EXCEED] <----

    if cmd == CMD_REMOVE:
      .....

    return result

govint commented:
Agree, the code can be corrrected to handle the incorrect input better. Added a new error code and message and fixed existing issues with check_privileges_for_command().

Tested with the new changes:

docker volume create -d vsphere --name vol3 -o size=10geeb33
Error response from daemon: create vol3: VolumeDriver.Create: Invalid volume size specified.
root@sc-rdops-vm02-dhcp-52-237:/mnt/docker-volume-vsphere# docker volume create -d vsphere --name vol3 -o size=10emmbee
Error response from daemon: create vol3: VolumeDriver.Create: Invalid volume size specified.
root@sc-rdops-vm02-dhcp-52-237:/mnt/docker-volume-vsphere# docker volume create -d vsphere --name vol3 -o size=10e
Error response from daemon: create vol3: VolumeDriver.Create: Invalid volume size specified.
root@sc-rdops-vm02-dhcp-52-237:/mnt/docker-volume-vsphere# docker volume create -d vsphere --name vol3 -o size=10
Error response from daemon: create vol3: VolumeDriver.Create: Invalid volume size specified.
root@sc-rdops-vm02-dhcp-52-237:/mnt/docker-volume-vsphere# docker volume create -d vsphere --name vol3 -o size=10mbi
Error response from daemon: create vol3: VolumeDriver.Create: Invalid volume size specified.
root@sc-rdops-vm02-dhcp-52-237:/mnt/docker-volume-vsphere# docker volume create -d vsphere --name vol3 -o size=10gbi
Error response from daemon: create vol3: VolumeDriver.Create: Invalid volume size specified.

govint commented:
CI failed because of #1395

pshahzeb commented:
This PR failed due to Failed to invoke command [docker volume rm clone_src_volume_59anldo9p9od]: exit status 1

----------------------------------------------------------------------
FAIL: volumecreate_test.go:49: VolumeCreateTestSuite.TearDownTest

volumecreate_test.go:54:
    c.Assert(err, IsNil, Commentf(out))
... value *exec.ExitError = &exec.ExitError{ProcessState:(*os.ProcessState)(0xc4203e9600), Stderr:[]uint8(nil)} ("exit status 1")
... Error response from daemon: unable to remove volume: remove clone_src_volume_59anldo9p9od: VolumeDriver.Remove: Failed to remove volume clone_src_volume_59anldo9p9od, in use by VM = photon.ga.vmfs.


----------------------------------------------------------------------
PANIC: volumecreate_test.go:176: VolumeCreateTestSuite.TestValidOptions

volumecreate_test.go:183:
    c.Assert(err, IsNil, Commentf(out))
... value *exec.ExitError = &exec.ExitError{ProcessState:(*os.ProcessState)(0xc420317500), Stderr:[]uint8(nil)} ("exit status 1")
... Error response from daemon: create clone_src_volume_59anldo9p9od: Post http://%!F(MISSING)run%!F(MISSING)docker%!F(MISSING)plugins%!F(MISSING)d2833c6789e74a581ba316e51419b0df4b87f156fc32de8544cc7905b15431dd%!F(MISSING)vsphere.sock/VolumeDriver.Create: http: ContentLength=51 with Body length 0

... Panic: Fixture has panicked (see related PANIC)
2017/06/13 09:54:19 Finding VM name from IP Address [192.168.31.241]
2017/06/13 09:54:19 Finding VM name from IP Address [192.168.31.163]



Although not related to this PR.
Should we merge this PR?

govint commented:
WE can merge but @lipingxue has asked for CI to pass. Any test thats showing errors should be disabled so CI isn't getting the errors.

pshahzeb commented:
Agreed. The failure isn't due to the changes this PR in my view. I think the CI failed due to plugin restart (as per the logs i can see)  and we already have a issue filed for this.

govint commented:
CI failed with issue in #1371

govint commented:
Have added debug code to check if the restart error can be identified. Will revert the debug changes later today and update the PR.

pdhamdhere commented:
2 approvals, green CI. Merging.
